<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Codon optimization</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Codon optimization</h1>



<div id="introduction" class="section level3">
<h3>Introduction</h3>
<p>Codon optimization modifies a gene’s coding sequence to enhance
protein production without changing the encoded amino acid sequence. The
<code>codon_optimize</code> function in the <code>cubar</code> package
provides three strategies for optimizing coding sequences based on the
codon usage of the target organism. The first two strategies replace
rare codons with more frequently used ones, while the third strategy
employs the third-party deep learning model
<code>CodonTransformer</code> (Fallahpour et al., 2025) to optimize
codon usage.</p>
</div>
<div id="evironment-setup" class="section level3">
<h3>Evironment setup</h3>
<p>Additionally, <code>cubar</code> can integrate the state-of-the-art
deep learning model <code>SpliceAI</code> (Jaganathan et al., 2019) to
prevent the unintended introduction of cryptic splice sites in optimized
sequences. As both <code>SpliceAI</code> and
<code>CodonTransformer</code> are Python-based, users must manually
install these packages. Here we demonstrates how to install them using
<code>conda</code> (or <code>mamba</code>) in a new environment for use
in <code>cubar</code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># create a new environment named &quot;cubar_env&quot; with both python and r installed</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-n</span> cubar_env python=3.12 r-base blas=<span class="pp">*</span>=netlib r-reticulate</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co"># activate the environment we just created</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="ex">conda</span> activate cubar_env</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co"># install CodonTransformer and SpliceAI</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="ex">pip</span> install CodonTransformer tensorflow spliceai</span></code></pre></div>
</div>
<div id="optimization-strategies" class="section level3">
<h3>Optimization strategies</h3>
<div id="naive-method" class="section level5">
<h5>“naive” method</h5>
<p>The default “naive” method simply replaces each codon to the most
preferred one in the same family or subfamily.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">library</span>(cubar)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>seq <span class="ot">&lt;-</span> <span class="st">&#39;ATGCTACGA&#39;</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>cf_all <span class="ot">&lt;-</span> <span class="fu">count_codons</span>(yeast_cds)</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co">#&gt; Loading required package: Biostrings</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;Biostrings&#39; was built under R version 4.3.3</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;BiocGenerics&#39;</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:stats&#39;:</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="co">#&gt;     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co">#&gt;     as.data.frame, basename, cbind, colnames, dirname, do.call,</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co">#&gt;     duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co">#&gt;     lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="co">#&gt;     pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="co">#&gt;     tapply, union, unique, unsplit, which.max, which.min</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="co">#&gt; Loading required package: stats4</span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;S4Vectors&#39;</span></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a><span class="co">#&gt; The following object is masked from &#39;package:utils&#39;:</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a><span class="co">#&gt;     findMatches</span></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb2-29"><a href="#cb2-29" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-30"><a href="#cb2-30" tabindex="-1"></a><span class="co">#&gt;     I, expand.grid, unname</span></span>
<span id="cb2-31"><a href="#cb2-31" tabindex="-1"></a><span class="co">#&gt; Loading required package: IRanges</span></span>
<span id="cb2-32"><a href="#cb2-32" tabindex="-1"></a><span class="co">#&gt; Loading required package: XVector</span></span>
<span id="cb2-33"><a href="#cb2-33" tabindex="-1"></a><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span id="cb2-34"><a href="#cb2-34" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;GenomeInfoDb&#39; was built under R version 4.3.3</span></span>
<span id="cb2-35"><a href="#cb2-35" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-36"><a href="#cb2-36" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;Biostrings&#39;</span></span>
<span id="cb2-37"><a href="#cb2-37" tabindex="-1"></a><span class="co">#&gt; The following object is masked from &#39;package:base&#39;:</span></span>
<span id="cb2-38"><a href="#cb2-38" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-39"><a href="#cb2-39" tabindex="-1"></a><span class="co">#&gt;     strsplit</span></span>
<span id="cb2-40"><a href="#cb2-40" tabindex="-1"></a>optimal_codons <span class="ot">&lt;-</span> <span class="fu">est_optimal_codons</span>(cf_all)</span>
<span id="cb2-41"><a href="#cb2-41" tabindex="-1"></a>seq_opt <span class="ot">&lt;-</span> <span class="fu">codon_optimize</span>(seq, optimal_codons)</span>
<span id="cb2-42"><a href="#cb2-42" tabindex="-1"></a><span class="fu">print</span>(seq_opt)</span>
<span id="cb2-43"><a href="#cb2-43" tabindex="-1"></a><span class="co">#&gt; 9-letter DNAString object</span></span>
<span id="cb2-44"><a href="#cb2-44" tabindex="-1"></a><span class="co">#&gt; seq: ATGCTACGT</span></span></code></pre></div>
</div>
<div id="idt-method" class="section level5">
<h5>“IDT” method</h5>
<p>The “IDT” option implements the method used by the codon optimization
tool of Integrated DNA Technologies. Briefly, this method randomly
selects synonymous codons from the same family or subfamily based on
their relative frequency, but excluding rare codons used below 10% in
the target organism.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>seq_opt <span class="ot">&lt;-</span> <span class="fu">codon_optimize</span>(seq, <span class="at">cf =</span> cf_all, <span class="at">method =</span> <span class="st">&quot;IDT&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">print</span>(seq_opt)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="co">#&gt; 9-letter DNAString object</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">#&gt; seq: ATGCTCCGC</span></span></code></pre></div>
</div>
<div id="codontransformer-method" class="section level5">
<h5>“CodonTransformer” method</h5>
<p>The “CodonTransformer” method optimizes codon usage with the
third-party software <code>CodonTransformer</code> directly using a
wrapper in R. <code>CodonTransformer</code> is a deep learning model
that can generate coding sequences that show similar codon usage and
distribution to host genes with reduced negative cis elements in a wide
range of organisms across the tree of life. Please refer to the original
study for more details.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>seq_opt <span class="ot">&lt;-</span> <span class="fu">codon_optimize</span>(seq, <span class="at">method =</span> <span class="st">&quot;CodonTransformer&quot;</span>, <span class="at">organism =</span> <span class="st">&quot;Saccharomyces cerevisiae&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">print</span>(seq_opt)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#&gt; 9-letter DNAString object</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt; seq: ATGTTAAGATGA</span></span></code></pre></div>
<p><code>cubar</code> can generate several optimized sequences at the
same time using the argument <code>num_sequences</code> with the method
“IDT” and “CodonTransformer”. When <code>num_sequences</code> is greater
than 1, identical duplicate sequences will be retained as a single copy,
potentially resulting in a final sequence count less than the specified
value.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>seqs_opt <span class="ot">&lt;-</span> <span class="fu">codon_optimize</span>(seq, <span class="at">cf =</span> cf_all, <span class="at">method =</span> <span class="st">&quot;IDT&quot;</span>, <span class="at">num_sequences =</span> <span class="dv">10</span>)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">print</span>(seqs_opt)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co">#&gt; DNAStringSet object of length 6:</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co">#&gt;     width seq</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#&gt; [1]     9 ATGCTCCGT</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co">#&gt; [2]     9 ATGCTGCGT</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co">#&gt; [3]     9 ATGCTTCGT</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co">#&gt; [4]     9 ATGCTACGT</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt; [5]     9 ATGCTCCGA</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt; [6]     9 ATGCTTCGA</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>seqs_opt <span class="ot">&lt;-</span> <span class="fu">codon_optimize</span>(seq, <span class="at">method =</span> <span class="st">&quot;CodonTransformer&quot;</span>, <span class="at">organism =</span> <span class="st">&quot;Saccharomyces cerevisiae&quot;</span>,</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="at">num_sequences =</span> <span class="dv">10</span>, <span class="at">deterministic =</span><span class="cn">FALSE</span>, <span class="at">temperature =</span> <span class="fl">0.4</span>)</span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="fu">print</span>(seqs_opt)</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="co">#&gt; DNAStringSet object of length 4:</span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co">#&gt;     width seq</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="co">#&gt; [1]    12 ATGTTGAGATAA</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="co">#&gt; [2]    12 ATGTTAAGATAA</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="co">#&gt; [3]    12 ATGTTGAGATGA</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a><span class="co">#&gt; [4]    12 ATGTTGAGATAG</span></span></code></pre></div>
</div>
</div>
<div id="splice-site-detection" class="section level3">
<h3>Splice site detection</h3>
<p>In addition, <code>cubar</code> integrated the deep learning tool
<code>SpliceAI</code> to identify potential splice sites with the
argument <code>spliceai</code>. When the probabilities of non-splice
site for each base are greater than 0.5, it is considered that there are
no potential splice junction sites, and the
<code>Possible_splice_junction</code> in the output is marked as FALSE,
otherwise it is marked as TRUE.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>seqs_opt <span class="ot">&lt;-</span> <span class="fu">codon_optimize</span>(seq, <span class="at">cf =</span> cf_all, <span class="at">method =</span> <span class="st">&quot;IDT&quot;</span>, <span class="at">num_sequences =</span> <span class="dv">10</span>, <span class="at">spliceai =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">print</span>(seqs_opt)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#&gt;     Candidate_optimized_sequence Possible_splice_junction</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co">#&gt;                           &lt;char&gt;                   &lt;lgcl&gt;</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt;  1:                    ATGCTACGC                    FALSE</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt;  2:                    ATGCTGCGA                    FALSE</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt;  3:                    ATGCTGCGT                    FALSE</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt;  4:                    ATGCTTCGC                    FALSE</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co">#&gt;  5:                    ATGCTACGT                    FALSE</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co">#&gt;  6:                    ATGCTTCGG                    FALSE</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co">#&gt;  7:                    ATGCTCCGT                    FALSE</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co">#&gt;  8:                    ATGCTTCGT                    FALSE</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="co">#&gt;  9:                    ATGCTCCGA                    FALSE</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="co">#&gt; 10:                    ATGCTTCGA                    FALSE</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a>seq_opt <span class="ot">&lt;-</span> <span class="fu">codon_optimize</span>(seq, <span class="at">method =</span> <span class="st">&quot;CodonTransformer&quot;</span>, <span class="at">organism =</span> <span class="st">&quot;Saccharomyces cerevisiae&quot;</span>, <span class="at">spliceai =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="fu">print</span>(seq_opt)</span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="co">#&gt;    Candidate_optimized_sequence Possible_splice_junction</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="co">#&gt;                          &lt;char&gt;                   &lt;lgcl&gt;</span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="co">#&gt; 1:                 ATGTTAAGATGA                    FALSE</span></span></code></pre></div>
</div>
<div id="recommendations" class="section level3">
<h3>Recommendations</h3>
<p>Codon usage within a coding sequence influences multiple aspects of
mRNA biology, including RNA secondary structure, translation elongation,
co-translational folding, and mRNA stability (Liu et al., 2021).
Achieving an optimal coding sequence requires carefully balancing all of
these factors, making codon optimization a more complex task than it may
initially appear. The <code>naive</code> approach implemented in
cubar—which simply replaces each non-optimal codon with the most
preferred one—fails to account for local sequence context and may lead
to unintended consequences. For instance, rare codons are often
strategically positioned to facilitate proper folding of protein
domains; indiscriminately replacing them with preferred codons can
disrupt this process and promote aggregation (Moss et al., 2024).</p>
<p>Given these complexities, we recommend using more sophisticated and
context-aware methods for codon optimization. Established strategies
such as that implemented in the IDT’s codon optimization tool have
demonstrated long-term effectiveness and are widely adopted within the
research community. Meanwhile, newer approaches based on deep generative
models—such as CodonTransformer (Fallahpour et al., 2025) and CodonBert
(Li et al., 2024)—leverage large-scale natural sequence data and
advanced architectures like recurrent neural networks and attention
mechanisms to capture context-dependent codon usage patterns (Novakovsky
et al., 2023). These deep learning–based models offer a powerful,
flexible framework for codon optimization and are likely to play a
central role in future sequence design workflows.</p>
<div id="references" class="section level5">
<h5>References</h5>
<ul>
<li>Fallahpour A, Gureghian V, Filion GJ, Lindner AB, Pandi A.
CodonTransformer: a multispecies codon optimizer using context-aware
neural networks. Nat Commun. 2025 Apr 3;16(1):3205. doi:
10.1038/s41467-025-58588-7. PMID: 40180930; PMCID: PMC11968976.</li>
<li>Jaganathan K, Kyriazopoulou Panagiotopoulou S, McRae JF, Darbandi
SF, Knowles D, Li YI, Kosmicki JA, Arbelaez J, Cui W, Schwartz GB, Chow
ED, Kanterakis E, Gao H, Kia A, Batzoglou S, Sanders SJ, Farh KK.
Predicting Splicing from Primary Sequence with Deep Learning. Cell. 2019
Jan 24;176(3):535-548.e24. doi: 10.1016/j.cell.2018.12.015. Epub 2019
Jan 17. PMID: 30661751.</li>
<li>Method used by the IDT codon optimization tool: <a href="https://sg.idtdna.com/pages/education/decoded/article/idt-codon-optimization-tool-makes-synthetic-gene-design-easy" class="uri">https://sg.idtdna.com/pages/education/decoded/article/idt-codon-optimization-tool-makes-synthetic-gene-design-easy</a></li>
<li>Liu Y, Yang Q, Zhao F. Synonymous but Not Silent: The Codon Usage
Code for Gene Expression and Protein Folding. Annu Rev Biochem. 2021 Jun
20;90:375-401. doi: 10.1146/annurev-biochem-071320-112701. Epub 2021 Jan
13. PMID: 33441035; PMCID: PMC8284178.</li>
<li>Moss MJ, Chamness LM, Clark PL. The Effects of Codon Usage on
Protein Structure and Folding. Annu Rev Biophys. 2024 Jul;53(1):87-108.
doi: 10.1146/annurev-biophys-030722-020555. Epub 2024 Jun 28. PMID:
38134335; PMCID: PMC11227313.</li>
<li>Li S, Moayedpour S, Li R, Bailey M, Riahi S, Kogler-Anele L, Miladi
M, Miner J, Pertuy F, Zheng D, Wang J, Balsubramani A, Tran K, Zacharia
M, Wu M, Gu X, Clinton R, Asquith C, Skaleski J, Boeglin L, Chivukula S,
Dias A, Strugnell T, Montoya FU, Agarwal V, Bar-Joseph Z, Jager S.
CodonBERT large language model for mRNA vaccines. Genome Res. 2024 Aug
20;34(7):1027-1035. doi: 10.1101/gr.278870.123. PMID: 38951026; PMCID:
PMC11368176.</li>
<li>Novakovsky G, Dexter N, Libbrecht MW, Wasserman WW, Mostafavi S.
Obtaining genetics insights from deep learning via explainable
artificial intelligence. Nat Rev Genet. 2023 Feb;24(2):125-137. doi:
10.1038/s41576-022-00532-2. Epub 2022 Oct 3. PMID: 36192604.</li>
</ul>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
